azure gateway: return MD5Sum as ETag for S3 API compatibility

poornas opened the pull request:
Fixes #6872.

<!--- Provide a general summary of your changes in the Title above -->

ETag generated in PutObjectPart is a UUID string rather than the MD5Sum of data being uploaded for Azure gateway. This poses a problem when edge caching mode is on, because part ETag validation is done at the time of CompleteMultipartUpload and it is expected that ETag be a hex encoded string.
## Description
<!--- Describe your changes in detail -->
Since the Azure BlockBlob api does not allow setting metadata, save a temporary part metadata file with the association between the uuid based ETag (used as part of the block id) and the md5sum which is known after data is written to backend. This temporary part metadata is cleaned up on completion of upload or aborting of upload. This allows PutObjectPart to be [S3 compatible](https://docs.aws.amazon.com/AmazonS3/latest/API/RESTCommonResponseHeaders.html) w.r.t etag
## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
allows objects uploaded to Azure via multipart API to be saved in edge cache when edge cache is enabled.
## Regression
<!-- Is this PR fixing a regression? (Yes / No) --> No
<!-- If Yes, optionally please include minio version or commit id or PR# that caused this regression, if you have these details. -->

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
1. enable edge cache ```export MINIO_CACHE_DRIVES=/tmp/testcache```
2. start azure gateway ./minio gateway azure
3. mc cp /tmp/large myazure/testbucket/multipart
4.  You should see testbucket/multipart cached in /tmp/testcache folder
## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added unit tests to cover my changes.
- [ ] I have added/updated functional tests in [mint](https://github.com/minio/mint). (If yes, add `mint` PR # here: )
- [x] All new and existing tests passed.

krishnasrinivas commented on cmd/gateway/azure/gateway-azure.go:
```diff
@@ -1141,10 +1190,33 @@ func (a *azureObjects) CompleteMultipartUpload(ctx context.Context, bucket, obje
+			return objInfo, minio.InvalidPart{}
+		}
+		defer func() {
+			derr := pblob.Delete(nil)
```
should be deleted only in case of successful CompleteMultipartUpload()

krishnasrinivas commented on cmd/gateway/azure/gateway-azure.go:
```diff
@@ -46,12 +46,14 @@ import (
+	azureBlockSize                 = 100 * humanize.MiByte
+	azureS3MinPartSize             = 5 * humanize.MiByte
+	metadataObjectNameTemplate     = minio.GatewayMinioSysTmp + "multipart/v1/%s.%x/azure.json"
+	metadataObjectPartNameTemplate = minio.GatewayMinioSysTmp + "multipart/v1/%s.%x.%x/azure.json"
```
We were encoding all the information into blockIDs itself so that we don't have to store an external metadata file like this. But now if we are having a part-meta file for each part then we don't have to encode anything in blockIDs and instead encode all the info in this meta file itself. This way blockIDs are just UUIDs and the part-meta file will look like this:

metadataObjectPartNameTemplate = minio.GatewayMinioSysTmp + "multipart/v1/%s.%x.%x/part-%d.json"
```
{
    etag: "d9cca721a735dac4efe709e0f3518373",
    blockIDs: ["4ef280b9-5752-49f8-ab6c-bc464c7dde0e", "0beb2b5a-633e-410a-a43e-6725bc366ef4", ....]
}
```

Also the following naming conventions will make it harder for future maintainers:
```
	metadata["x-amz-meta-etag"] = etag
	metadata["x-amz-meta-md5sum"] = md5Hex
```

i.e here x-amz-meta-etag is the unique string we use for azure and x-amz-meta-md5sum is the etag that is returned to s3-client. We can clean it up if we make the above change. Let's discuss with @abperiasamy

poornas replied:
@krishnasrinivas, refactored to use the per part metadata file as per discussion for multipart operation.

poornas committed:
azure gateway: return MD5Sum as ETag for S3 API compatibility
Fixes #6872.

This PR refactors multipart upload implementation to use a per
part metadata file which is cleaned up at the end of the upload

krishnasrinivas commented on cmd/gateway/azure/gateway-azure.go:
```diff
@@ -261,6 +264,30 @@ func s3MetaToAzureProperties(ctx context.Context, s3Metadata map[string]string)
+	PartID   string   `json:"partID"`
+	Size     int64    `json:"Size"`
+	BlockIDs []string `json:"blockIDs"`
+	MD5Sum   string   `json:"md5sum"`
```
can be renamed to ETag, since it is actually ETag and we using MD5sum as ETag

poornas replied:
ok, renamed this to ETag

harshavardhana reviewed (dismissed):
LGTM tested
